core/src/main/java/org/elasticsearch/index/query/QueryShardContext.java

```diff
@@ -343,38 +331,38 @@ public ParsedQuery parse(XContentParser parser) {
      */
     @Nullable
     public ParsedQuery parseInnerFilter(XContentParser parser) throws IOException {
-        reset(parser);
+        reset();
+        parser.setParseFieldMatcher(this.parseFieldMatcher());
         try {
-            parseFieldMatcher(indexSettings.getParseFieldMatcher());
-            Query filter = QueryBuilder.rewriteQuery(parseContext().parseInnerQueryBuilder(), this).toFilter(this);
+            Query filter = QueryBuilder.rewriteQuery(newParseContext(parser).parseInnerQueryBuilder(), this).toFilter(this);
             if (filter == null) {
                 return null;
             }
             return new ParsedQuery(filter, copyNamedQueries());
         } finally {
-            reset(null);
+            reset();
         }
     }
 
 
     private ParsedQuery innerParse(XContentParser parser) throws IOException, QueryShardException {
-        reset(parser);
+        reset();
+        parser.setParseFieldMatcher(this.parseFieldMatcher());
         try {
-            parseFieldMatcher(indexSettings.getParseFieldMatcher());
-            Query query = parseInnerQuery();
+            Query query = parseInnerQuery(parser);
             return new ParsedQuery(query, copyNamedQueries());
         } finally {
-            reset(null);
+            reset();
         }
     }
 
-    public Query parseInnerQuery() throws IOException {
-        return toQuery(this.parseContext().parseInnerQueryBuilder(), this);
+    public Query parseInnerQuery(XContentParser parser) throws IOException {
+        parser.setParseFieldMatcher(this.parseFieldMatcher());
```

Comment:
it is odd that we do the opposite here, as the context holds the parse field matcher and sets it to the parser. But I do see why. The parse field matcher is used for more than just parsing xcontent. I think it's ok then.